### PR TITLE
Info updates

### PR DESCRIPTION
**Pull request by danilkolikov**
Added updating of post's info

- Services
- Exponential backoff
- Database
- Rx

### CODE REVIEW

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    @Inject
+    @SuppressWarnings("NullableProblems")   // onCreate
+    @NonNull
+    CoreModel coreModel;
+
+    @NonNull
+    private final Map<Long, Subscription> updateIntervals;
+
+    @Nullable
+    private rx.Subscription updateSubscription;
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: prefix rx?

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        Utils.getApplicationContext(this).getAppComponent().inject(this);
+    }
+
+    @Nullable
+    @Override
+    public IBinder onBind(Intent intent) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: nullability

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        return result;
+    }
+
+    /**
+     * Drop post's subscription for updates
+     *
+     * @param loudlyPost Post to remove subscription
+     * @return True, if subscription was removed, false otherwise
+     */
+    public boolean unSubscribe(@NonNull LoudlyPost loudlyPost) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: maybe unsubscribe?

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        return subscribed;
+    }
+
+    /**
+     * Drop post's subscription for updates
+     *
+     * @param loudlyPosts List of posts for drop subscription
+     * @return True, if any subscription was dropped
+     */
+    public boolean unSubscribe(@NonNull List<LoudlyPost> loudlyPosts) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: here too

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        return ids;
+    }
+
+    @CheckResult
+    @NonNull
+    private Observable<Info> getUpdates(@NonNull SolidList<LoudlyPost> posts) {
+        return coreModel
+                .observeConnectedNetworksModels()
+                .<Pair<SinglePost, Info>>flatMap(networkContract -> {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: why do you need generics here?

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        if (loudlyPost == null) {
+            return Observable.empty();
+        }
+        return postsDatabaseModel.updateStoredInfo(loudlyPost, post.getNetwork(), diff);
+    }
+
+    private long getRefreshInterval() {
+        long minimal = MAXIMAL_UPDATE_INTERVAL;
+        for (Subscription subscription : updateIntervals.values()) {
+            minimal = Math.min(minimal, subscription.interval);
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: static import) maybe

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    @CheckResult
+    @NonNull
+    public Single<Boolean> subscribeOnFrequentUpdates(@NonNull List<LoudlyPost> loudlyPosts) {
+        return getService()
+                .map(service -> service.subscribe(loudlyPosts, MINIMAL_UPDATE_INTERVAL));
+    }
+
+    @CheckResult
+    @NonNull
+    public Single<Boolean> unSubscribe(@NonNull LoudlyPost loudlyPost) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: unsubscribe

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    @UiThread
+    @CheckResult
+    @NonNull
+    private Single<UpdateInfoService.UpdateInfoServiceBinder> bindToService() {
+        return Observable
+                .<UpdateInfoService.UpdateInfoServiceBinder>create(observer -> {
+                    Intent startService = new Intent(loudlyApplication, UpdateInfoService.class);
+                    loudlyApplication.bindService(startService, new ServiceConnection() {
+                        @Override
+                        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: nullability

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                            UpdateInfoService.UpdateInfoServiceBinder binder =
+                                    (UpdateInfoService.UpdateInfoServiceBinder) iBinder;
+                            serviceBinder = binder;
+                            observer.onNext(binder);
+                            observer.onCompleted();
+                        }
+
+                        @Override
+                        public void onServiceDisconnected(ComponentName componentName) {
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: and here

File: app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java [hunk 10]
```diff
@@ -615,15 +654,17 @@ private PlainPost fromStored(@NonNull StoredPost storedPost,
...
+    public Observable<SolidList<LoudlyPost>> loadPostsByTimeInterval(@NonNull TimeInterval timeInterval) {
         if (cached.isEmpty()) {
             return selectPostsByTimeInterval(timeInterval)
                     .map(list -> {
                         cached.addAll(list);
                         return asSolidList(list);
                     });
         } else {
-            final DividedList<PlainPost> dividedList =
+            final DividedList<LoudlyPost> dividedList =
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java]: loudly posts?

File: app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java [hunk 11]
```diff
@@ -644,7 +685,295 @@ private PlainPost fromStored(@NonNull StoredPost storedPost,
...
+                ))
+                .prepare()
+                .asRxObservable()
+                .map(this::countInfo);
+    }
+
+    @CheckResult
+    @NonNull
+    public Observable<Info> updateStoredInfo(@NonNull LoudlyPost loudlyPost,
+                                             @Networks.Network int network,
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java]: static import

File: app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java [hunk 12]
```diff
@@ -644,7 +685,295 @@ private PlainPost fromStored(@NonNull StoredPost storedPost,
...
+                        }
+                    }
+                    return true;
+                });
+    }
+
+    @CheckResult
+    @NonNull
+    private Single<Boolean> deleteOldestStoredEvent(@NonNull LoudlyPost post,
+                                                    @Event.EventType short type,
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java]: static import

File: app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java [hunk 13]
```diff
@@ -644,7 +685,295 @@ private PlainPost fromStored(@NonNull StoredPost storedPost,
...
+                ))
+                .prepare()
+                .asRxSingle()
+                .flatMap(this::deleteStoredEvents);
+    }
+
+    @CheckResult
+    @NonNull
+    private Single<Boolean> updateEvents(@NonNull LoudlyPost post,
+                                         @Networks.Network int network,
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java]: and here

File: app/src/main/java/ly/loud/loudly/base/entities/Event.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    @NonNull
+    public Info toInfo() {
+        switch (type) {
+            case LIKE:
+                return new Info(1, 0, 0);
+            case SHARE:
+                return new Info(0, 1, 0);
+            default:    // Comment
+                return new Info(0, 0, 1);
+        }
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/base/entities/Event.java]: what is this?

File: app/src/main/java/ly/loud/loudly/util/RxUtils.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+     * Infinitely resubscribe to this Observable after specified amount of time
+     *
+     * @param observable       An observable to resubscribe
+     * @param intervalSupplier Supplier of interval for updates
+     * @param <T>              Type of elements in observable
+     * @return Infinitely repeated observable, emitted with pauses between them
+     */
+    @CheckResult
+    @NonNull
+    public static <T> Observable<T> repeat(@NonNull Observable<T> observable,
```
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/util/RxUtils.java]: norm)

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 1 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: We have other Subscription here

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 2 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: Default method

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 3 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: Idea said that there is no such word

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 5 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: Trouble with type inference

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 6 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: nope

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java (hunk 8 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: default method

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java (hunk 9 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: and here

File: app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java (hunk 10 above)
Comment (danilkolikov) [inline app/src/main/java/ly/loud/loudly/application/models/PostsDatabaseModel.java]: For avoiding instanceof checks

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 1 above)
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: ok

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 2 above)
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: also need nullability

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 3 above)
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: lets discuss

File: app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java (hunk 6 above)
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/UpdateInfoService.java]: (

File: app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java (hunk 8 above)
Comment (ZeRoGerc) [inline app/src/main/java/ly/loud/loudly/application/models/InfoUpdateModel.java]: also needs nullability

**Comment by ZeRoGerc** [comment]
ok, with issues

**Comment by danilkolikov** [comment]
issues fix'd

**Pull request merged**